### DISCUSSION TITLE
Extremely slow performance when using mongodb

### DISCUSSION THREAD

**Discussion by ricksearcy** [elsa-workflows/elsa-core#2423] [General]
Using this example as a test https://github.com/elsa-workflows/elsa-guides/tree/master/src/ElsaGuides.ContentApproval.Web

With the default SqlLite persistence provider we are seeing normal 1000ms response times. 

Switching to  .MongoDb persistence the same requests are taking 50-60 seconds. We use mongo for our db so obviously this is an issue. Any ideas? This is a mongo server we use for our other app and it performs fine, so I don't believe it's related to the actual DB server. 

UseMongoDbPersistence(opts =>
                    {
                        opts.ConnectionString = “ourconnstring”;
                        opts.DatabaseName = “ourdb”;
                    })

**Comment by ricksearcy** [comment]
This is the culprit. I quit counting after it was hit more 50 times from one request to the /Documents endpoint. 

 public async Task<int> CountAsync(ISpecification<T> specification, CancellationToken cancellationToken = default) =>
            await Collection.AsQueryable().Apply(specification).CountAsync(cancellationToken);


Elsa.Persistence.MongoDb.dll!Elsa.Persistence.MongoDb.Stores.MongoDbStore<Elsa.Models.WorkflowInstance>.CountAsync(Elsa.Persistence.Specifications.ISpecification<Elsa.Models.WorkflowInstance> specification, System.Threading.CancellationToken cancellationToken) Line 74
	at C:\elsa-core-master (1)\elsa-core-master\src\persistence\Elsa.Persistence.MongoDb\Stores\MongoDbStore.cs(74)
Elsa.Core.dll!Elsa.Persistence.Decorators.EventPublishingWorkflowInstanceStore.CountAsync(Elsa.Persistence.Specifications.ISpecification<Elsa.Models.WorkflowInstance> specification, System.Threading.CancellationToken cancellationToken)
Elsa.Core.dll!Elsa.Decorators.CachingWorkflowRegistry.WorkflowHasUnfinishedWorkflowsAsync(Elsa.Services.Models.IWorkflowBlueprint workflowBlueprint, System.Threading.CancellationToken cancellationToken)
Elsa.Core.dll!Elsa.Decorators.CachingWorkflowRegistry.<ListActiveInternalAsync>d__12.MoveNext()
[Resuming Async Method]
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Elsa.Decorators.CachingWorkflowRegistry.<ListActiveInternalAsync>d__12>.MoveNext(System.Threading.Thread threadPoolThread)
System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask)
System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining)
System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject)

**Reply by sfmskywalker** [reply]
See response at https://github.com/elsa-workflows/elsa-core/issues/2433
